### PR TITLE
[MLIR] Enable GPU Dialect to SYCL runtime integration

### PR DESCRIPTION
**Pull request by silee2**
GPU Dialect lowering to SYCL runtime is driven by spirv.target_env attached to gpu.module. As a result of this, spirv.target_env remains as an input to LLVMIR Translation.
A SPIRVToLLVMIRTranslation without any actual translation is added to avoid an unregistered error in mlir-cpu-runner.
SelectObjectAttr.cpp is updated to
1) Pass binary size argument to getModuleLoadFn
2) Pass parameter count to getKernelLaunchFn
This change does not impact CUDA and ROCM usage since both mlir_cuda_runtime and mlir_rocm_runtime are already updated to accept and ignore the extra arguments.

### CODE REVIEW

**Comment by silee2** [comment]
This is last part of refactoring #65539

File: mlir/test/Integration/GPU/SYCL/gpu-to-spirv.mlir [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// RUN: mlir-opt %s -pass-pipeline='builtin.module(spirv-attach-target{ver=v1.0 caps=Addresses,Int64,Kernel},convert-gpu-to-spirv{use-64bit-index=true},gpu.module(spirv.module(spirv-lower-abi-attrs,spirv-update-vce)),func.func(llvm-request-c-wrappers),convert-scf-to-cf,convert-cf-to-llvm,convert-arith-to-llvm,convert-math-to-llvm,convert-func-to-llvm,gpu-to-llvm{use-bare-pointers-for-kernels=true},gpu-module-to-binary,expand-strided-metadata,lower-affine,finalize-memref-to-llvm,reconcile-unrealized-casts)' \
```
Comment (fabianmcg) [inline mlir/test/Integration/GPU/SYCL/gpu-to-spirv.mlir:1]: Can we get a couple more integration tests?

File: mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
+//===- SPIRVToLLVMIRTranslation.cpp - Translate SPIRV to LLVM IR ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a translation between the MLIR SPIRV dialect and
```
Comment (fabianmcg) [inline mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp:9]: `SPIRV` -> `SPIR-V`

File: mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp [hunk 3]
```diff
@@ -377,10 +394,21 @@ llvm::LaunchKernel::createKernelLaunch(mlir::gpu::LaunchFuncOp op,
   if (!binary)
     return op.emitError() << "Couldn't find the binary: " << binaryIdentifier;
 
+  llvm::Constant *paramsCount =
+      llvm::ConstantInt::get(i64Ty, op.getNumKernelOperands());
+
+  std::string binarySizeIdentifier = getBinarySizeIdentifier(moduleName);
+  Value *binarySizeVar = module.getGlobalVariable(binarySizeIdentifier, true);
```
Comment (fabianmcg) [inline mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp:401]: Why can't you use the initializer in `Value *binary = module.getGlobalVariable(binaryIdentifier, true);` to obtain the size of the binary?

Comment (silee2) [reply mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp]: Done.

File: mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp (hunk 2 above)
Comment (silee2) [reply mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp]: Done.

File: mlir/test/Integration/GPU/SYCL/gpu-to-spirv.mlir (hunk 1 above)
Comment (silee2) [reply mlir/test/Integration/GPU/SYCL/gpu-to-spirv.mlir]: Added two more tests.

File: mlir/include/mlir/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+#ifndef MLIR_TARGET_LLVMIR_DIALECT_SPIRV_SPIRVTOLLVMIRTRANSLATION_H
+#define MLIR_TARGET_LLVMIR_DIALECT_SPIRV_SPIRVTOLLVMIRTRANSLATION_H
+
+namespace mlir {
+
+class DialectRegistry;
+class MLIRContext;
+
+/// Register the SPIRV dialect and the translation from it to the LLVM IR in the
```
[the remark this answers is not in the document]
Comment (silee2) [reply mlir/include/mlir/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.h]: Done.

File: mlir/test/Target/LLVMIR/gpu.mlir [hunk 5]
```diff
@@ -4,6 +4,7 @@
 module attributes {gpu.container_module} {
   // CHECK: [[ARGS_TY:%.*]] = type { i32, i32 }
   // CHECK: @kernel_module_bin_cst = internal constant [4 x i8] c"BLOB", align 8
+  // CHECK:  @kernel_module_bin_size_cst = internal constant i64 4, align 8
```
Comment (fabianmcg) [inline mlir/test/Target/LLVMIR/gpu.mlir:7]: Please update the tests with the new changes.

File: mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp [hunk 6]
```diff
@@ -377,10 +379,17 @@ llvm::LaunchKernel::createKernelLaunch(mlir::gpu::LaunchFuncOp op,
   if (!binary)
     return op.emitError() << "Couldn't find the binary: " << binaryIdentifier;
 
+  auto binaryVar = dyn_cast<llvm::GlobalVariable>(binary);
+  llvm::Constant *binaryInit = binaryVar->getInitializer();
+  auto binaryDataSeq = dyn_cast<llvm::ConstantDataSequential>(binaryInit);
```
Comment (fabianmcg) [inline mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp:384]: Some form of check must be placed on the pointers:
```suggestion
  auto binaryVar = dyn_cast<llvm::GlobalVariable>(binary);
  assert(binaryVar && "expected a global variable");
  llvm::Constant *binaryInit = binaryVar->getInitializer();
  auto binaryDataSeq = dyn_cast_or_null<llvm::ConstantDataSequential>(binaryInit);
  assert(binaryDataSeq && "expected a valid initializer");
```

File: mlir/test/Integration/GPU/SYCL/gpu-addf32-to-spirv.mlir [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  func.func @main() {
+    %0 = memref.get_global @__constant_2x2x2xf32 : memref<2x2x2xf32>
+    %1 = memref.get_global @__constant_2x2x2xf32_0 : memref<2x2x2xf32>
+    %2 = call @test(%0, %1) : (memref<2x2x2xf32>, memref<2x2x2xf32>) -> memref<2x2x2xf32>
+    %cast = memref.cast %2 : memref<2x2x2xf32> to memref<*xf32>
+    call @printMemrefF32(%cast) : (memref<*xf32>) -> ()
+    return
+  }
+  func.func private @printMemrefF32(memref<*xf32>)
+  func.func @test(%arg0: memref<2x2x2xf32>, %arg1: memref<2x2x2xf32>) -> memref<2x2x2xf32> {
```
Comment (Dinistro) [inline mlir/test/Integration/GPU/SYCL/gpu-addf32-to-spirv.mlir:20]: Ultra nit drive by comment: Could you indent the body of this function as you did for the other function bodies?

File: mlir/test/Target/LLVMIR/gpu.mlir (hunk 5 above)
Comment (silee2) [reply mlir/test/Target/LLVMIR/gpu.mlir]: Done.

File: mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp (hunk 6 above)
Comment (silee2) [reply mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp]: Added pointer checks.

File: mlir/test/Integration/GPU/SYCL/gpu-addf32-to-spirv.mlir (hunk 7 above)
Comment (silee2) [reply mlir/test/Integration/GPU/SYCL/gpu-addf32-to-spirv.mlir:20]: Done.

File: mlir/include/mlir/Target/LLVMIR/Dialect/All.h [hunk 8]
```diff
@@ -61,6 +63,7 @@ registerAllGPUToLLVMIRTranslations(DialectRegistry &registry) {
   registerLLVMDialectTranslation(registry);
   registerNVVMDialectTranslation(registry);
   registerROCDLDialectTranslation(registry);
+  registerSPIRVDialectTranslation(registry);
```
[the remark this answers is not in the document]
Comment (silee2) [reply mlir/include/mlir/Target/LLVMIR/Dialect/All.h:66]: @joker-eph Any thoughts on library vs inlining the call?

**Comment by silee2** [comment]
@antiagainst @kuhar Any comments?

File: mlir/include/mlir/Target/LLVMIR/Dialect/All.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/include/mlir/Target/LLVMIR/Dialect/All.h:66]: Trying to understand...
What is the unit test that requires this?

File: mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+#include "mlir/Dialect/SPIRV/IR/SPIRVDialect.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Target/LLVMIR/ModuleTranslation.h"
+
+using namespace mlir;
+using namespace mlir::LLVM;
+
+void mlir::registerSPIRVDialectTranslation(DialectRegistry &registry) {
+  registry.insert<spirv::SPIRVDialect>();
```
Comment (joker-eph) [inline mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp:24]: I'm confused here: there is no translation interface.

Comment (silee2) [reply mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp:24]: There is no translation but dialect registration is still required as spirv.target_env is attached and appears in input.
Other option is to register SPIR-V dialect directly here: mlir/include/mlir/Target/LLVMIR/Dialect/All.h
@joker-eph Any thoughts? Or better option?

File: mlir/include/mlir/Target/LLVMIR/Dialect/All.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/include/mlir/Target/LLVMIR/Dialect/All.h:66]: Which muir-translate test requires that? I don't find it by skimming the patch.

**Review by antiagainst — APPROVED** [review]
LGTM regarding SPIR-V. But please address comments from others. :)

File: mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp (hunk 9 above)
Comment (silee2) [reply mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp:24]: @joker-eph

**Comment by silee2** [comment]
@fabianmcg Can you approve and merge? @joker-eph is not responding and this PR has been open for a long time.
It is blocking other future PRs that depend on this one.
If SPIR-V dialect registration needs to change that can be done in another PR.

File: mlir/include/mlir/Target/LLVMIR/Dialect/All.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (silee2) [reply mlir/include/mlir/Target/LLVMIR/Dialect/All.h:66]: @fabianmcg Added test.
Any other request? The PR has been open for a while and blocking other work that depends on it. Would be great if you can approve soon.

**Review by fabianmcg — APPROVED** [review]
LGTM! I'll squash and merge as soon it passes all pre check tests.

File: mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp (hunk 9 above)
Comment (joker-eph) [reply mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp:24]: It does not make sense to me to have an API called "registerSPIRVDialectTranslation" that itself registers a dialect and no translation actually.

File: mlir/include/mlir/Target/LLVMIR/Dialect/All.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/include/mlir/Target/LLVMIR/Dialect/All.h:66]: > The PR has been open for a while and blocking other work that depends on it. Would be great if you can approve soon.

Well, we were waiting on the test that you just added today!

File: mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp (hunk 9 above)
Comment (fabianmcg) [reply mlir/lib/Target/LLVMIR/Dialect/SPIRV/SPIRVToLLVMIRTranslation.cpp:24]: I'm sorry I missed these. I had the same concern, the alternative was adding an inline registration call. However, I then realized that adding `#include "mlir/Dialect/SPIRV/IR/SPIRVDialect.h"` to `mlir/Target/LLVMIR/Dialect/All.h` was needed , which seemed even more undesirable thus I changed my mind and agreed on the current scheme. 
One way to clean it, is removing inline function from `mlir/Target/LLVMIR/Dialect/All.h` and creating a `MLIRToLLVMIR` library.

**Pull request merged**